---
title: "Analyzing distribution of ChIPseq reads across clusters"
output: html_document
---
```{r setup, echo=FALSE}
knitr::opts_knit$set(root.dir = '../EXPERIMENTS/9_ChIPseq_2_Clstr')
knitr::opts_chunk$set(fig.path='Figures/')
knitr::opts_chunk$set(message=FALSE,warning=FALSE,error=FALSE)
```
A comparison of the relative abundance of the clusters between the genome assemblies and the input (K=30, cluster assignment) highlights clear discrepancies. In comparison to the Input, clusters 4 and 1 are under-represented in the assemblies, while clusters 6,5 and 2 are over-represented.  
```{r Cluster_Abundance, echo=FALSE}
library(reshape)
library(ggplot2)
library(phyclust)
source("./Scripts/prop_plotting_functions.R")
seqdata <- read.fasta("q_CEN2_All_K6.aln2")
ClstrT10 <- sapply(seqdata$seqname[grepl("Chr",seqdata$seqname)],parse_clstr)
ClstrT10.p <-prop.table(table(ClstrT10))
ClstrPB <- sapply(seqdata$seqname[!grepl("Chr",seqdata$seqname)],parse_clstr)
ClstrPB.p <- prop.table(table(ClstrPB))
NChIP.032015 <- read.csv("./2015-03-NChIP-K30/2015-03-NChIP-K30_results.csv", header=T)
Input <-NChIP.032015[which(NChIP.032015$filename == "SML_14_merged.fa" & NChIP.032015$Class == "Unique"),]
Input <-Input[-c(1,2)]
colnames(Input) <-c("1","2","3","4","5","6")
ClstrInput.p <- prop.table(t(Input))
prop.df <- cbind(ClstrT10.p,ClstrPB.p,ClstrInput.p)
colnames(prop.df) <- c("TAIR10","PacBio \nContigs","INPUT")
prop.df <- as.data.frame(prop.df)
prop.df$Clusters <- as.factor(rownames(prop.df))
prop.df.melted <- melt(prop.df)

custom <-c(heat.colors(10)[3],terrain.colors(10)[c(1,4)],heat.colors(10)[7],topo.colors(10)[c(4,3)])
p <-ggplot(prop.df.melted, aes(x=variable,y=value,fill=Clusters)) + geom_bar(stat="identity")
p <- p + scale_fill_manual(values=custom) + theme_bw()
p <- p + theme(axis.title=element_blank(),axis.text=element_text(size=12))
p
```

Fraction of ChIPseq reads that originate from CEN repeats. This estimate is based on binning reads into two fractions, those that generate Kmers that overlap with a set of signature CEN kmers extracted from the above clusters.
```{r ChIPseq_fractions, echo=FALSE}
K25.032015 <- read.csv("./2015-03-NChIP-K25/2015-03-NChIP-K25_summary.csv", header=T)
K25.082015 <- read.csv("./2015-08-NChIP-K25/2015-08-NChIP-K25_summary.csv", header=T)
K25.summary <- rbind(K25.032015,K25.082015)
K25.out <-cen.fraction(K25.summary,"ChIP_metadata.csv")
K25.out.plot <-plot.cen.fraction(K25.out)

K30.032015 <- read.csv("./2015-03-NChIP-K30/2015-03-NChIP-K30_summary.csv", header=T)
K30.082015 <- read.csv("./2015-08-NChIP-K30/2015-08-NChIP-K30_summary.csv", header=T)
K30.summary <- rbind(K30.032015,K30.082015)
K30.out <-cen.fraction(K30.summary,"ChIP_metadata.csv")
K30.out.plot <-plot.cen.fraction(K30.out)

K25.out.plot
K30.out.plot
```

Next, I asked what is the distribution across the six clusters of reads in a ChIPseq library that originate from CEN repeats. It seems clear that all histone marks tested, except for *A. thaliana* CENH3 and *L. oleraceum* CENH3 have the same distribution as in the input, i.e. their distribution across the clusters follows their relative abundance in the genome. However, both the native and divergent CENH3 show a distinct enrichment for CEN repeats from Cluster 4 and a depletion for repeats from Cluster 3.
```{r Cluster_distribution, echo=FALSE, message=FALSE, warning=FALSE}
K30.032015 <- read.csv("./2015-03-NChIP-K30/2015-03-NChIP-K30_results.csv", header=T)
K30.082015 <- read.csv("./2015-08-NChIP-K30/2015-08-NChIP-K30_results.csv", header=T)
K30.results <- rbind(K30.032015,K30.082015)
K30.out.1 <-clstr.distribution.1(K30.results,"ChIP_metadata.csv","Unique")
K30.out.plot.1 <-plot.clstr.distribution(K30.out.1)
K30.out.2 <-clstr.distribution.2(K30.results,"ChIP_metadata.csv","Unique")
K30.out.plot.2 <-plot.clstr.distribution(K30.out.2)

K30.out.plot.1
K30.out.plot.2
```

To identify potential chromosome-specific repeats, I visualized the phylogenetic relationship of repeats extracted from the TAIR10 genome assembly and coloured the tips by their chromosome of origin.
```{r, echo=FALSE,eval=FALSE}
library(ape)
library(phyclust)
seqdata <- read.fasta("q_CEN2_All_K6.aln2")
mat <- tolower(seqdata$org.code)
rownames(mat) <- seqdata$seqname
DNAbin <- as.DNAbin(mat)
TAIR10seqs <- DNAbin[grepl("Chr",labels(DNAbin)),]
TAIR10tree <- nj(dist.dna(TAIR10seqs,pairwise.deletion=TRUE,model="F84"))
save(TAIR10tree,file="./RData/TAIR10tree.RData")
```
```{r Repeats_by_Chromosome, echo=FALSE}
load("./RData/TAIR10tree.RData")
n.tips <- length(TAIR10tree$tip.label)
col.vector <- vector(mode="character",length=nrow(TAIR10tree$edge))
col.vector[TAIR10tree$edge[,2]>n.tips] <- ":nonterminal---"
for(i in 1:n.tips){
    col.vector[which(TAIR10tree$edge[,2] == i)] <- TAIR10tree$tip.label[i]
}
Chr <- sapply(col.vector,parse_Chr)
Chr <- as.factor(Chr)
Cluster <- sapply(col.vector,parse_clstr)
Cluster <- as.factor(Cluster)

par(mar=c(0,0,0,0))
mat <-matrix(c(1,2,3,4),1,4,byrow=TRUE)
layout(mat,widths=c(1,0.4,1,0.4),heights=1)
custom <-brewer.pal(11,"Spectral")[c(9,7,4,3,2)]
custom <-c(gray(0),custom)
palette(custom)
plot(TAIR10tree, type="u",no.margin=TRUE,show.tip.label=FALSE, edge.color=Chr)
plot.new()
legend('left',levels(Chr)[-1], fill = custom[-1], bty="n",cex=1.2,x.intersp=0.2)
custom <-c(gray(0),heat.colors(10)[3],terrain.colors(10)[c(1,4)],heat.colors(10)[7],topo.colors(10)[c(4,3)])
palette(custom)
plot(TAIR10tree, type="u",no.margin=TRUE, show.tip.label=FALSE, edge.color=Cluster)
plot.new()
legend("left",legend=c("Cluster1","Cluster2","Cluster3","Cluster4","Cluster5","Cluster6"), fill = custom[-1], bty="n",cex=1.2,x.intersp=0.2)
```